src/controllers/midi/midisourceclock.h

```diff
@@ -0,0 +1,114 @@
+#ifndef MIDISOURCECLOCK_H
+#define MIDISOURCECLOCK_H
+
+#include <QList>
+
+#include "util/time.h"
+
+// Stub out the Time::elapsed function for easier testing.
+class MockableClock {
+  public:
+    virtual ~MockableClock() { };
+    virtual qint64 now() = 0;
+};
+
+class MixxxClock : public MockableClock {
+  public:
+    virtual qint64 now() {
+        return Time::elapsed();
+    }
+};
+
+// MidiSourceClock is not thread-safe, but is thread compatible using ControlObjects.
+// The MIDI thread will make calls into MidiSourceClock and then update two Control
+// Objects with the current reported BPM and last beat time.  The engine thread
+// can use those values to call a static function to calculate beat percentage
+// at any time in the future.  Time values are in nanoseconds for compatibility
+// with Time::elapsed().
+class MidiSourceClock {
+  public:
+    // The number of midi ticks per quarter note (1 beat in 4/4 time).
```

Comment:
I think you mean 1 beat per 4/4 measure. 

A 4/4 measure is a special case. Can't we just use kPulsesPerBeat? 
What is the difference between pulses and ticks? If it is the same, please decide for one.